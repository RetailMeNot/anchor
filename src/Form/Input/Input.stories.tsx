// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
// VENDOR
import styled from 'styled-components';
// COMPONENT
import Input from './Input.component';
import {Icon} from '../../Icon';

const StyledStory = styled.div`
  padding: 2rem 5rem;
`;

const validate = (value: any) => {
  if (value !== 'test') {
    return 'You can only input test';
  }
  return null;
};

const formatPhoneNumber = (value: any) => {
  let phoneNumber = '';
  const input = value.replace(/\D/g, '').substring(0, 10);
  const zip = input.substring(0, 3);
  const middle = input.substring(3, 6);
  const last = input.substring(6, 10);
  if (input.length > 6) {
    phoneNumber = `${zip}-${middle}-${last}`;
  } else if (input.length > 3) {
    phoneNumber = `${zip}-${middle}`;
  } else if (input.length >= 0) {
    phoneNumber = `${zip}`;
  }
  return phoneNumber;
};

storiesOf('Components/Input', module)
  .add('with Prefix', () => (
    <StyledStory>
      <Input
        label="Search"
        prefix={<Icon type="search"/>}
      />
    </StyledStory>
  ))
  .add('without Prefix', () => (
    <StyledStory>
      <Input label="First Name" />
    </StyledStory>
  ))
  .add('with errors', () => (
    <StyledStory>
      <Input label="Validated Input" validate={validate} />
    </StyledStory>
  ))
  .add('with format', () => (
    <StyledStory>
      <Input label="Phone Number" format={formatPhoneNumber} />
    </StyledStory>
  ))
  .add('with placeholder', () => (
    <StyledStory>
      <Input label="Email" placeholder="email@rmn.com" />
    </StyledStory>
  ))
  .add('disabled', () => (
    <StyledStory>
      <Input label="Disabled Example" disabled={true} />
    </StyledStory>
  ));
