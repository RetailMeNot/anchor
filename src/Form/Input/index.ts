export { default as Input } from './Input.component';
