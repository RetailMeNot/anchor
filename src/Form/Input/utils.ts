export const INPUT_KEY = 'inputs';
export const INPUT_THEME = {
    sizes: {
        sm: {
            height: '2.5rem',

            input: {
                height: '1rem',
                fontSize: '0.875rem',
                lineHeight: '1rem',
            },
        },
        md: {
            height: '3.125rem',
            padding: '0.25rem',

            input: {
                height: '1rem',
                fontSize: '0.875rem',
                lineHeight: '1rem',
            },
        },
        lg: {
            height: '3.125rem',
            padding: '0.25rem',

            input: {
                height: '1.5rem',
                fontSize: '1rem',
                lineHeight: '1.5rem',
            },
        },
    },
};
