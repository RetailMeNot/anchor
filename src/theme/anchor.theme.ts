// ANCHOR THEME
import { radii } from './sizes.theme';
import { MODAL_THEME_KEY, MODAL_THEME } from '../Modal';

// Copied from Adapter.StandardBreakpoints mins
// todo: dedup- they should use a single source
const breakpoints = {
    sm: '576px',
    md: '768px',
    lg: '992px',
    xl: '1200px',
};

interface AnchorTheme {
    colors?: any;
    breakpoints?: any;
    radii?: any;
    [key: string]: any;
}

export const anchorTheme: AnchorTheme = {
    breakpoints,
    radii,
    [MODAL_THEME_KEY]: MODAL_THEME,
};
