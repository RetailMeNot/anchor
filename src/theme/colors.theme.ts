export default {
  neutrals: {
    charcoal: {
      base: '#323232',
      light: '#595959',
      dark: '#000000',
    },
    ash: {
      base: '#A6A6A6',
      light: '#D3D3D3',
      dark: '#808080',
    },
    silver: {
      base: '#F1F1F1',
      light: '#FAFAFA',
      dark: '#E7E7E7',
    },
    lavenderWhispers: {
      base: '#F7F5FC',
      light: '#FBFAFF',
      dark: '#F1EEFA',
    },
    white: {
      base: '#FFFFFF',
    },
  },
  main: {
    grapePurchase: {
      base: '#602D6C',
      light: '#91609C',
      dark: '#4B2354',
    },
    rx: {
      base: '#F23466',
      light: '#FF6B9C',
      dark: '#CC2D52',
    },
    genie: {
      base: '#4C18F3',
      light: '#7855D9',
      dark: '#42199A',
    },
  },
  accent: {
    savvyCyan: {
      base: '#0998D6',
      light: '#66CCFF',
      dark: '#0074A6',
    },
    tealBreaker: {
      base: '#00CCCD',
      light: '#00E0D1',
      dark: '#00BDBD',
    },
  },
  secondary: {
    dealEnvy: {
      base: '#14A382',
      light: '#00CDAD',
      dark: '#268068',
    },
    flashPink: {
      base: '#DF206C',
      light: '#E665A1',
      dark: '#B81D5B',
    },
    cyberMango: {
      base: '#EB663D',
      light: '#FA9F5A',
      dark: '#C74E30',
    },
    goldMine: {
      base: '#F5A623',
      light: '#FFC761',
      dark: '#CC861D',
    },
    fireSale: {
      base: '#EA4545',
      light: '#F16667',
      dark: '#CC3B3B',
    },
  },
  alert: {
    error: {
      base: '#CC3B3B',
    },
    success: {
      base: '#268068',
    },
  },
  rx: {
    purple: {
      base: '#A678B5',
    },
  },
};
