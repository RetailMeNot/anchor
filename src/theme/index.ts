export { default as colors } from './colors.theme';
export { default as fonts } from './fonts.theme';
export { default as sizes } from './sizes.theme';
export { default as variables } from './variables.theme';

export { NormalizeCSS, GlobalCSS } from './GlobalStyles';
