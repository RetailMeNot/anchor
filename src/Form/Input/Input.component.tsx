// REACT
import * as React from 'react';
// VENDOR
import classNames from 'classnames';
import styled from 'styled-components';
// THEME
import { fonts, colors } from '../../theme';

interface InputProps {
  className?: string;
  label?: string;
  placeholder?: string;
  disabled?: boolean;
  defaultValue?: any;
  value?: any;
  // onFocus?: () => void;
  // onBlur?: () => void;
  onChange?: (e: any) => void;
  validate?: (e: any) => void;
  format?: (e: any) => void;

  prefix?: any;
  suffix?: any;
}

interface InputState {
  isFocused: boolean;
  value: string;
  errorMessage: string;
}

export const StyledInput = styled.div `
  width: calc(100% - 2rem);
  height: 3rem;
  position: relative;
  display: flex;
  flex-direction: column;
  font-family: ${fonts.fontFamily};
  opacity: ${({disabled}: InputProps) => disabled && '0.5'};

  .field-input {
    width: 100%;
    box-sizing : border-box;
    padding: 24px 0 6px 0;
    padding-left: ${({ prefix }: InputProps) => !prefix ? '.75rem' : 'calc(3rem - 3px)'};
    border-radius: 2px;
    border: solid 1px ${colors.neutrals.ash.light};
    position: relative;
    font-size: calc(1rem - 2px);
    font-weight: 500;
    background: none;
    z-index: 1;

    &:active,
    &:focus {
      outline: none;
      border-color: ${({ disabled }: InputProps) => disabled ? colors.neutrals.ash.light : colors.neutrals.ash.dark};
    }

    &.error-border {
      border-color: ${colors.alert.error.base};
    }
  }

  .input-prefix {
    padding-top: 1rem;
    padding-left: 1rem;
    position: absolute;
  }

  .field-label {
    position: absolute;
    margin-left: ${({ prefix }: InputProps) => !prefix ? '.75rem' : 'calc(3rem - 3px)'};
    margin-top: calc(1rem - 1px);
    top: 0;
    font-size: calc(1rem - 2px);
    font-weight: 500;
    line-height: 1.29;
    transform-origin: 0% 50%;
    transition: all 200ms ease-in-out;
    color: ${colors.neutrals.charcoal.light};
    cursor: text;
  }

  .field-input:focus + .field-input:not([value='']) {
    z-index: 2;
    opacity: 1;
  }

  .field-input:focus + .field-label,
  .field-input:not([value='']) + .field-label {
    z-index: 2;
    transform: scale(0.80) translateY(-.75rem);
    opacity: 0.75;
    cursor: default;
  }

  .error {
    margin-top: .5rem;
    font-size: .75rem;
    font-weight: 500;
    color: ${colors.alert.error.base};
  }
`;

export class Input extends React.Component<InputProps> {
  readonly state: InputState = {
    isFocused: false,
    value: '',
    errorMessage: '',
  };

  render(): React.ReactElement<Input> {
    const {
      label,
      prefix,
      disabled,
    } = this.props;

    const {
      value,
      isFocused,
      errorMessage,
    } = this.state;

    return (
      <StyledInput prefix={prefix} disabled={disabled}>
        {prefix && React.cloneElement(prefix, {
          height: 16,
          width: 16,
          color: colors.neutrals.ash.dark,
          className: 'input-prefix',
        })}
        <input
          disabled={disabled}
          onFocus={() => this.setState({ ...this.state, isFocused: true })}
          onBlur={() => this.setState({ ...this.state, isFocused: false })}
          onChange={(event: React.ChangeEvent<HTMLInputElement>) => this.handleChange(event)}
          value={value}
          className={classNames('field-input', { ['error-border']: errorMessage})}
        />
        <label className="field-label">
          {label}
        </label>
        {!isFocused && errorMessage && <div className="error">{errorMessage}</div>}
      </StyledInput>
    );
  }

  private handleChange(event: React.ChangeEvent<HTMLInputElement>): void {
    const errorMessage = this.props.validate && event.target.value && this.props.validate(event.target.value);
    const value = this.props.format ? this.props.format(event.target.value) : event.target.value;
    this.setState({
      value,
      errorMessage,
    });
  }
}

export default Input;
