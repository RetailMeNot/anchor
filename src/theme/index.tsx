// Theme Files
import { typography } from './typography.theme';
import { ColorsTheme } from './colors.theme';
import { ThemeProvider as XstyledThemeProvider } from '@xstyled/styled-components';
import styled from '@xstyled/styled-components';
import { space as spaceStyles, SpaceProps } from '@xstyled/system';
import { BUTTON_KEY, BUTTON_THEME } from '../Button';
import { INPUT_KEY, INPUT_THEME } from '../Form/Input/utils';
import React from 'react';
import {Button} from '../Button';
import classNames from 'classnames';

export const RootTheme = {
    typography,
    radii: { none: '0', base: '4px', modal: '8px', circular: '1000px' },
    [BUTTON_KEY]: BUTTON_THEME,
    [INPUT_KEY]: INPUT_THEME,
    colors: ColorsTheme,
    fonts: {
        base:
            'Avenir Next, Segoe UI, Helvetica Neue, Helvetica, Roboto, sans-serif',
        mono: `'SFMono-Regular', Consolas, Liberation Mono, Menlo, Courier, monospace`,
    },
    // These are all minimums
    breakpoints: {
        xs: 0,
        sm: 576,
        md: 768,
        lg: 992,
        xl: 1200,
        xxl: 1600,
    },
    // todo: we should likely remove the color.borders colors and just use this
    // or find a way to reference the colors.border colors here
    // e.g. `thin solid borders.base`
    borders: {
        light: `thin solid ${ColorsTheme.borders.light}`,
        base: `thin solid ${ColorsTheme.borders.base}`,
        dark: `thin solid ${ColorsTheme.borders.dark}`,
    },

    skeleton: {
        colorStart: '#E7E7E7',
        colorEnd: '#D3D3D3',
    },
};

// Exporting out @xstyled's ThemeProvider. Only way I could get it to work w/o issues.
// export class ThemeProvider2 extends XstyledThemeProvider {}

interface IterableObject {
    [s: number]: number | string | boolean | IterableObject;
  }

interface DebugProps {
    property: number | string | boolean | IterableObject;
    propertyName: string;
    rootProperty?: boolean;
}

interface StyledDebugProps {
    indent?: number;
}

const StyledDebug = styled('div')<StyledDebugProps>`
    font-family: mono;
    margin-top: 0.5rem;
    color: white;

    .category .category h2 {
        background-color: #ec8947;
    }

    .category .category .box {
        border-color: #ec8947;
    }

    .category .category .category h2 {
        background-color: #b84343;
        color: white;
    }

    .category .category .category .box {
        border-color: #b84343;
    }

    .category .category .category .category h2 {
        background-color: #91cdd2;
        color: black;
    }

    .category .category .category .category .box {
        border-color: #91cdd2;
    }

    .category .category .category .category .category h2 {
        background-color: #567224;
        color: white;
    }

    .category .category .category .category .category .box {
        border-color: #567224;
    }
`;

const StyledH2 = styled('h2')`
    ${spaceStyles}
    background-color: #ffea92;
    color: black;
    margin-bottom: 0;
    margin-top: 0;
    padding: 0.25rem 0.5rem 0.25rem 1rem;
    font-size: 1rem;
    cursor: pointer;
    border-radius: 0 1rem 1rem 0;
`;

interface BoxProps extends SpaceProps {
    show?: boolean;
}

const Box = styled('div')<BoxProps>`
    ${spaceStyles}
    width: 20rem;
    border-left: solid #ffea92 2px;
    margin: 0.5rem;
    ${({show}) => `display: ${show ? 'block' : 'none'}`}
`;

const Category = styled('div')<SpaceProps>`
    ${spaceStyles}
    width: 20rem;
`;


const ColorBox = styled('div')<any>`
    width: 1rem;
    height: 1rem;
    display: inline-block;
    vertical-align: middle;
    margin-left: 0.5rem;
    background-color: ${({backgroundColor}) => backgroundColor};
    border: solid white 1px;
`;

const Debug = ({property, propertyName}: DebugProps) => {
    const [show, setShow] = React.useState(false);

    return(
        <StyledDebug>
            {(property || property === 0)
                ? (
                    typeof property === 'number' ||
                    typeof property === 'string' ||
                    typeof property === 'boolean' ||
                    typeof property === 'function'
                        ? (
                                <>
                                    <span style={{fontWeight: 'bold'}}>{propertyName}: </span>
                                    <span>
                                        {typeof property === 'function'
                                            ? 'function'
                                            : /^#[a-zA-Z0-9]{4,7}/.test(property.toString())
                                                ? (<>
                                                    <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                                    <ColorBox backgroundColor={property} />
                                                    </>)
                                                : <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                        }
                                    </span>
                                </>
                            )
                        :
                            <Category pl={0} className="category">
                                <StyledH2 onClick={() => setShow(!show)}>{propertyName}</StyledH2>
                                <Box pl={2} className="box" show={show}>
                                    {Object.values(property).map((_property, index) => (
                                        <Debug
                                            key={index}
                                            property={_property}
                                            propertyName={Object.getOwnPropertyNames(property)[index]}
                                        />
                                    ))}
                                </Box>
                            </Category>
                    )
                : (
                    <>
                       Empty.
                    </>
                )
            }
        </StyledDebug>
    );
};

const DebugWrapper = styled('div')`
    background-color: rgba(0,0,0,.75);
    border-left-color: transparent;
    position: fixed;
    left: 0;
    top: 0;
    height: 100%;
    overflow: hidden;
    width: 30rem;
    padding: 1rem 2rem 1rem 0;

    &.closed {
        height:2.4rem;
        width: auto;
        padding-right: 1rem;

        .debug-scroll {
            display: none;
        }
    }
`;

const DebugScroll = styled('div')`
    height: 100%;
    overflow-y: auto;
    width: 100%;
`;

interface ThemeProviderProps {
    children?: any;
    debug?: boolean;
    theme?: any;
}

export const ThemeProvider = ({ children, debug, theme }: ThemeProviderProps) => {
    const isOpen = localStorage.themeProviderDebug === 'true';
    const [isOpenState, setIsOpenState] = React.useState(isOpen);

    return (
        <XstyledThemeProvider theme={theme}>
            {debug &&
                <DebugWrapper className={classNames(!isOpenState ? 'closed' : null)}>
                    <Button
                        onClick={() => {
                            localStorage.themeProviderDebug = !isOpenState;
                            setIsOpenState(!isOpenState);
                        }}
                        size="sm"
                    >
                        ThemeProvider Debug
                    </Button>

                    <DebugScroll className="debug-scroll">
                        <Debug property={theme} propertyName="Theme" />
                    </DebugScroll>
                </DebugWrapper>
            }
            {children}
        </XstyledThemeProvider>
    );
};
