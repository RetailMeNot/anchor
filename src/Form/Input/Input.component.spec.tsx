// VENDOR
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { ThemeProvider } from '@xstyled/styled-components';
// COMPONENT
import { Input } from './Input.component';
import { RootTheme } from '../../theme';
import {
    ArrowBack as ArrowBackIcon,
    ArrowForward as ArrowForwardIcon,
} from '../../Icon';

describe('Component: Form/Input', () => {
    it('should be defined', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <Input />
            </ThemeProvider>
        );
        const wrapper = mount(subject);
        const component = shallow(subject);

        expect(subject).toBeDefined();
        expect(wrapper).toBeDefined();
        expect(component).toBeDefined();
        const tree = renderer.create(subject).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should pass additional input attributes via inputProps', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <Input inputProps={{ autoCapitalize: 'false' }} />
            </ThemeProvider>
        );
        const wrapper = mount(subject);

        expect(wrapper.find({ autoCapitalize: 'false' })).toBeDefined();
    });

    it('should fire passed functions when events occur', () => {
        let value = 0;
        const changeValue = () => {
            value++;
        };
        const changeMock = jest.fn();

        const subject = (
            <ThemeProvider theme={RootTheme}>
                <Input
                    onKeyDown={changeValue}
                    onKeyUp={changeValue}
                    onBlur={changeValue}
                    onFocus={changeValue}
                    onChange={changeMock}
                    value="test"
                />
            </ThemeProvider>
        );
        const wrapper = mount(subject);
        const input = wrapper.find('input');

        input.simulate('keyDown');
        expect(value).toBe(1);
        input.simulate('keyUp');
        expect(value).toBe(2);
        input.simulate('blur');
        expect(value).toBe(3);
        input.simulate('focus');
        expect(value).toBe(4);
        input.simulate('change');
        expect(changeMock.mock.calls[0][0]).toBe('test');
    });

    it('should render suffix and prefix JSX.', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <Input
                    prefix={<ArrowForwardIcon />}
                    suffix={<ArrowBackIcon />}
                />
            </ThemeProvider>
        );
        const wrapper = mount(subject);

        expect(wrapper.find('anchor-icon.arrow-back')).toBeTruthy();
        expect(wrapper.find('anchor-icon.arrow-forward')).toBeTruthy();
    });
});
