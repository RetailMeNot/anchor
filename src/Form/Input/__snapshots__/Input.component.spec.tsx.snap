// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Form/Input should be defined 1`] = `
.c0 {
  display: block;
  position: relative;
  border: solid thin #D3D3D3;
  border-radius: 4px;
  cursor: text;
  box-sizing: border-box;
  min-width: 5rem;
  width: 100%;
  margin: 0;
  overflow: hidden;
  height: 3.125rem;
  padding: 0.25rem;
}

.c0::-webkit-input-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c0::-moz-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c0:-ms-input-placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c0::placeholder {
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #808080;
}

.c0.focus {
  border-color: #808080;
}

.c0 label {
  -webkit-transform-origin: left bottom;
  -ms-transform-origin: left bottom;
  transform-origin: left bottom;
  height: 1.4rem;
  -webkit-transform: translate(0,0.6rem) scale(1);
  -ms-transform: translate(0,0.6rem) scale(1);
  transform: translate(0,0.6rem) scale(1);
}

.c0 label.lift {
  -webkit-transform: translate(0,0) scale(1);
  -ms-transform: translate(0,0) scale(1);
  transform: translate(0,0) scale(1);
}

.c0 input {
  height: 1rem;
  font-size: 0.875rem;
}

.c2 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column-reverse;
  -ms-flex-flow: column-reverse;
  flex-flow: column-reverse;
  padding: 0 0.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  box-sizing: border-box;
  border: none;
  padding: 0;
  outline: none !important;
  touch-action: manipulation;
  -webkit-appearance: none;
  background-color: transparent;
  z-index: 1;
  color: #323232;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
}

.c3[type='number']::-webkit-inner-spin-button,
.c3[type='number']::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c3:placeholder-shown + label {
  cursor: text;
  max-width: 100%;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

<div
  className="c0 anchor-input"
  onClick={[Function]}
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <input
        className="c3"
        name="input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        type="text"
      />
    </div>
  </div>
</div>
`;
